Consider to add more extensions to the ITimer interface

sivukhin opened the pull request:
Popular metrics libraries (like App.Metrics) provides a bunch of extensions to simplify metrics reporting. Usually, there are two basic methods: one that returns `IDisposable` that report metric at the `Dispose` call and the other that allows the user to measure the duration of lambda from the arguments (https://www.app-metrics.io/getting-started/metric-types/timers/).

A brief investigation of `vostok.metrics` library capabilities don't show any possible alternatives to the extension of `ITimer` that simplifies the duration measurement of lambda expressions. Of course, I can surround expression with using block but in the case when the return value is used in another part of the function body I need to extract some temporary variables in the outer scope and this makes code messier.

This PR extends `ITimer` with a bunch of extensions that simplify the measurement of the duration of lambdas. I choose to overload the `Measure(...)` function signature for the sake of API simplicity. I also add overloads for async operations but maybe this is a bad choice for a generic purpose library (for some reason App.Metrics and Metrics.NET don't have such overloads). Async overloads don't have `...Async` suffix in the function name because in this case the user of the library can misspell function call and get wrong measurements (in the case when sync version of the method was called).

sivukhin committed:
Add timer extensions

kungasc commented on Vostok.Metrics/Primitives/Timer/ITimerExtensions_Measurement.cs:
```diff
@@ -14,6 +15,58 @@ public static void Report([NotNull] this ITimer timer, TimeSpan value)
 
+        public static T Measure<T>([NotNull] this ITimer timer, Func<T> action)
+        {
+            var stopwatch = Stopwatch.StartNew();
```
наверно тут лучше просто обернуть в старый метод
```
using (timer.Measure())
  action()
```

kungasc commented on Vostok.Metrics/Primitives/Timer/ITimerExtensions_Measurement.cs:
```diff
@@ -14,6 +15,58 @@ public static void Report([NotNull] this ITimer timer, TimeSpan value)
         public static IDisposable Measure([NotNull] this ITimer timer)
             => new Measurement(timer);
 
+        public static T Measure<T>([NotNull] this ITimer timer, Func<T> action)
```
```suggestion
        public static T Measure<T>([NotNull] this ITimer timer, Func<T> func)
```

kungasc commented on Vostok.Metrics/Primitives/Timer/ITimerExtensions_Measurement.cs:
```diff
@@ -14,6 +15,58 @@ public static void Report([NotNull] this ITimer timer, TimeSpan value)
+            }
+        }
+
+        public static async Task<T> Measure<T>([NotNull] this ITimer timer, Func<Task<T>> action)
```
> Async overloads don't have ...Async suffix in the function name because in this case the user of the library can misspell function call and get wrong measurements (in the case when sync version of the method was called).

я бы наверно всё же добавил суффикс, ведь точно так же ты можешь забыть написать 

**await** Measure(...)

sivukhin replied:
Меня пугает, что ты можешь написать `var result = await timer.Measure(async () => await Do())` и код будет работать корректно, но метрика при этом измеряться не будет. Пока что мне кажется, что следование конвенции только создает лишние проблемы и ничем не помогает.

sivukhin replied:
Ага, видимо проблема с общими перегрузками в том, что без явной типизации у лямбды () => throw new Exception() выводится тип `Func<Task>` как более специфичный и вызывается не та перегрузка...((

sivukhin replied:
Разделил перегрузки, но у синхронной добавил явный суффикс `MeasureSync`.

sivukhin committed:
Adjust new test assertions for the sake of stability

kungasc commented on Vostok.Metrics/Primitives/Timer/ITimerExtensions_Measurement.cs:
```diff
@@ -14,6 +15,58 @@ public static void Report([NotNull] this ITimer timer, TimeSpan value)
+            }
+        }
+
+        public static async Task<T> Measure<T>([NotNull] this ITimer timer, Func<Task<T>> action)
```
```suggestion
        public static async Task<T> Measure<T>([NotNull] this ITimer timer, Func<Task<T>> func)
```

kungasc commented on Vostok.Metrics.Tests/Primitives/Timer/ITimerExtensions_Measurement_Tests.cs:
```diff
@@ -70,5 +71,59 @@ public void Measure_should_report_elapsed()
+            {
+                Thread.Sleep(0.1.Seconds());
+                throw new Exception();
+            });
```
добавить проверку, что Measure не съел Exception (ShouldThrow)

kungasc commented on Vostok.Metrics.Tests/Primitives/Timer/ITimerExtensions_Measurement_Tests.cs:
```diff
@@ -97,7 +98,8 @@ public void MeasureFunc_should_report_elapsed_and_return_value()
                 .Should()
                 .Be(42);
-            @event.Value.Should().BeApproximately(0.1, precision: 0.01);
+            @event.Value.Should().BeGreaterThan(0.1);
```
.Should().BeInRange()

kungasc replied:
но вообще, Thread.Sleep может и чуть меньше поспать)

sivukhin replied:
Уменьшил до 0.01s, но убрал проверку верхней границы, потому что тесты нестабильные в таком случае. Кажется для всех сценариев из этих тестов проверка верхней границы бесполезна.

kungasc replied:
да можно и просто != 0 проверить

sivukhin committed:
Review fixes

sivukhin committed:
Reduce sleep time for tests

sivukhin committed:
Loose assertion bounds for the sake of stability

- System clocks row resolution affect measurement and create possibility of flakiness
